packages/labs/compiler/src/lib/template-transform.ts

```diff
@@ -0,0 +1,337 @@
+      if (this.topLevelStatementToTemplate.has(node as ts.Statement)) {
+        return this.addTopLevelCompiledTemplate(node);
+      }
+
+      // Maybe rewrite the template expression to use a CompiledTemplateResult.
+      node = this.rewriteTemplateResultToCompiledTemplateResult(node);
+      return ts.visitEachChild(node, rewriteTemplates, this.context);
+    };
+
+    const rewrittenNode = ts.visitNode(node, rewriteTemplates);
+    if (!rewrittenNode) {
+      throw new Error(`Internal Error: Unexpected undefined 'rewrittenNode'.`);
+    }
+    return rewrittenNode;
+  }
+
+  /**
+   * This code is identical to the logic that occurs when preparing a lit-html
+   * `Template` (the internal class in lit-html.ts). It has been modified to use
+   * parse5 instead of the browser but otherwise is very similar.
+   *
+   * Because the prepared HTML will contribute to file size, markers have been
+   * stripped out, and comment nodes always use the 3 byte `<?>` format.
+   */
+  private litHtmlPrepareRenderPhase(templateExpression: ts.TemplateLiteral):
+    | {
+        shouldCompile: false;
+      }
+    | {shouldCompile: true; preparedHtml: string; parts: TemplatePart[]} {
+    const parts: Array<TemplatePart> = [];
+
+    // Get the static strings from the source code, turn them into a spoofed
+    // template strings array, and pass them into `getTemplateHtml` for the
+    // prepared HTML.
+    const spoofedTemplate = ts.isNoSubstitutionTemplateLiteral(
+      templateExpression
+    )
+      ? ([templateExpression.text] as unknown as TemplateStringsArray)
+      : ([
+          templateExpression.head.text,
+          ...templateExpression.templateSpans.map((s) => s.literal.text),
+        ] as unknown as TemplateStringsArray);
+    // lit-html enforces at runtime that `getTemplateHtml` only accepts a
+    // TemplateStringsResult.
+    (spoofedTemplate as unknown as {raw: string}).raw = '';
+    let html: TrustedHTML | string = '';
+    try {
+      [html] = getTemplateHtml(spoofedTemplate, 1);
+    } catch {
+      return {shouldCompile: false};
+    }
+    const ast = parseFragment(html as unknown as string, {
+      sourceCodeLocationInfo: false,
+    });
+
+    let nodeIndex = -1;
+    traverse(ast, {
+      'pre:node': (node) => {
+        if (isCommentNode(node)) {
+          if (node.data === markerMatch) {
+            parts.push({
+              type: PartType.CHILD,
+              index: nodeIndex,
+            });
+            // We have stored a reference to this comment node in the
+            // TemplatePart, so we can remove the comment text. At runtime we
+            // would leave the marker comment, but when compiling it's
+            // advantageous to save on file size.
+            node.data = '';
+          }
+        } else if (isTextNode(node)) {
+          // Do not count text nodes.
+          nodeIndex--;
+        }
+        nodeIndex++;
+      },
+    });
+
+    const preparedHtml = serialize(ast).replace(
```

Comment:
This is an interesting test case. I think this might be handled by https://github.com/lit/lit/pull/3984/files#diff-df7cea51b79a61536a1a04ef8f42d4a3f68ca366e8c0507d54b897d69041f128R279-R284 (which is coming after I incrementally add attribute parts). This handles raw text elements.

Reply:
That link isn't working for me for some reason, but I don't think that will be handled because the comment texts aren't bindings.

On the other hand, we could minimize down to `<?>` just the comments that we generate, looking for the marker text

Reply:
^ I like this approach. I'll roll this into a separate change.

Agree after thinking about it more that this would miscompile.

Reply:
Also added a TODO for this pending action.